Rename parameter selectors

ericpre opened the pull request:
## PR Summary

Follow up of #20558. Rename parameters (`maxdist`, `markers_props`, etc.) to make them consistent between the different selectors.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented on lib/matplotlib/widgets.py:
```diff
@@ -2436,10 +2438,19 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
             activated.
 
         marker_props : dict
+            Properties with which the interactive handles are drawn.  Currently
```
Better to just remove the entries that are no longer used.

tacaswell commented on lib/matplotlib/widgets.py:
```diff
@@ -2511,13 +2522,13 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
         else:
             props = dict(markeredgecolor=rectprops.get('edgecolor', 'r'))
-        props.update(cbook.normalize_kwargs(marker_props, Line2D._alias_map))
+        props.update(cbook.normalize_kwargs(handle_props, Line2D._alias_map))
```
This looks like they are used contrary to the docstring?

tacaswell reviewed (commented):
Seems reasonable, left 2 small comments.

ericpre commented:
Thanks for the review @tacaswell, it should be good now.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -2366,103 +2367,122 @@ def closest(self, x, y):
 
-                def onselect(eclick: MouseEvent, erelease: MouseEvent)
+    minspany : float, default: 0
+        Selections with an y-span less than *minspany* are ignored.
```
```suggestion
        Selections with a y-span less than *minspany* are ignored.
```

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -2511,13 +2531,13 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
         self._drawtype = drawtype
 
-        self.maxdist = maxdist
+        self.handle_grab_distance = handle_grab_distance
```
This needs a deprecation accessor on the old `maxdist` attribute.

ericpre replied:
Yes, indeed, done!

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -3033,15 +3068,15 @@ def __init__(self, ax, onselect, useblit=False,
 
         self._active_handle_idx = -1
-        self.vertex_select_radius = vertex_select_radius
+        self.handle_grab_distance = handle_grab_distance
```
This needs a deprecation accessor on the old `vertex_select_radius` attribute.

ericpre commented:
Should be all done!

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -2366,103 +2367,122 @@ def closest(self, x, y):
-        state_modifier_keys : dict, optional
-            Keyboard modifiers which affect the widget's behavior.  Values
-            amend the defaults.
+@docstring.Substitution(_RECTANGLESELECTOR_PARAMETERS_DOCSTRING)
```
Why this indirection?

ericpre replied:
To be able to reuse the docstring in EllipseSelector and RectangleSelector.

timhoffm replied:
I see. I'm not a fan of too much indirection in the docs. Some duplication is ok.
The risk with indirection is that people start to write about Rectangle explicitly, which does not make sense for the Ellipse case. Conversly, the risk with duplicated docstrings is that they are not in sync (which IMHO is not the end of the world).

It's a tradeoff which I would decide in favor of the duplication here. But I'm not going to fight over it.

ericpre replied:
I have put a placeholder for the artist type to be replaced by `ellipse` or `selector` to fix this issue.

> I even claim it's a mistake that `EllipseSelector` inherits from `RectangleSelector`.

I don't know about that, maybe for another PR! ;)

timhoffm reviewed (commented):
I know that some of the commented changes have precedence in other widget classes, however I'm not convinced, we should go that direction (e.g. code blocks for defaults).

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -2984,20 +3011,29 @@ class PolygonSelector(_SelectorWidget):
+
+            dict(color='k', linestyle='-', linewidth=2, alpha=0.5)
+
+    handle_props : dict
```
I see what you what you mean, but I am convinced that simplifying the API and making it more generic is worth it at the cost of the being slightly less precise. I agree that using a more specific terminology is more explicit but at the end of the day, when we create a selector - regardless of what selector it is - what we want is to set the properties of the patch and/or the handle, regardless of what their are or of their implementation details.
For example, as an average matplotlib user, I don't know what are the different parameters between `lineprops` and `markerprops`, I will need to look it up.

In 35ee92d, I have changed the various patch properties (`rectprops`, `lineprops`) argument to `props`, so that now we have only `props` and `handleprops` for all selectors.
To address #20618, we could add two method to set the properties of the selector patch and the handle separately. The idea is that either at initialisation or when using `set_props` or `set_handle_props`, we could set the properties regardless of what type of object they are.

I am happy to revert 35ee92d if others also think that this is going in the wrong direction!

timhoffm replied:
> For example, as an average matplotlib user, I don't know what are the different parameters between `lineprops` and `markerprops`, I will need to look it up.

Conversely, I argue that `props` is even less helpful. The different selectors are realized by different Artists and thus have different properties. We cannot hide that fact. `SpanSelector` accepts `facecolor`, `LassoSelector` accepts `solid_linestyle`, and neither accepts the other one respectively. With `props` the user has to lookup what is supported as well. But it's more surpising and harder to remember that the same parameter name accepts different inputs depending on the class.

ericpre commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,10 @@
+RectangleSelector and EllipseSelector
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``maxdist`` is deprecated, use ``handle_grab_distance`` argument instead
```
I am quite happy with `handle_grab_distance` that you introduced in #20558! :)

timhoffm replied:
I'm not. While it's not really bad, the `handle` part does not add much information / make it more specific (there's nothing else you could grab). `handle` just makes the name longer. Also, `range` is more appropriate than `distance` because it explicitly includes all values (otherwise the distance should rather be `max_distance`, but that's quite long and not better than `range`).

timhoffm commented:
Ok, so here is an overview of the current APIs

![grafik](https://user-images.githubusercontent.com/2836374/125193993-9aa91c80-e24f-11eb-90d9-8918a46525d2.png)


- Green entries are consistent (don't need any change).
- Orange entries should be synchronized.
- Yellow: needs discussion.

`props` and `handle_props` face the same problem that the selection indicator and the handles respectively are realized by different objects in the different classes and thus accept different values depending on the class. An additional complication is that `RectangleSelector` / `EllipseSelector` takes `lineprops` or `rectprops` depending on `drawtype`. Either solution for that is bad (`drawtype` deciding which parameters are accepted and `drawtype` deciding which values of a `prop` parameter are supported).


Given the overall situation I propose the following:

- `lineprops`/ `rectprops` get unified to `props`, but each docstring has to clearly specify which values are supported by linking the respective Artist properties.
- Similarly, we unifiy to `handle_props` and also link the Artist properties.
- We unify to `handle_grab_distance` / `vertex_select_radius` / `maxdist` to `grab_range` (or if you want `handle_grab_range`).

ericpre commented:
Thanks @timhoffm for the summary, this clarify the current situation!
I have improve the docstrings in 44b37fb, where I have tried to put the relevant links to the doc and I have rename `handle_grab_distance` to `grab_range` in 97456b4.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -1448,7 +1449,7 @@ class Cursor(AxesWidget):
     Other Parameters
     ----------------
-    **lineprops
+    **props
```
Let's keep this PR limited to the selectors. These are not artist properties that control a part of a selector. Instead, these are straight-forward properties of the centeral visual element. This is rather a case for #20048, which ``**props`` would move away from further. Note that there is also a substantial difference between a regular parameter with defaults and ``**kwargs``. You cannot assign to the latter explicitly, so that its name is not public API.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -1457,8 +1458,9 @@ class Cursor(AxesWidget):
+    @_api.rename_parameter("3.5", "lineprops", "props")
     def __init__(self, ax, horizOn=True, vertOn=True, useblit=False,
-                 **lineprops):
+                 **props):
```
As above: Not a selector, other scenario.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -1546,8 +1548,9 @@ class MultiCursor(Widget):
+    @_api.rename_parameter("3.5", "lineprops", "props")
     def __init__(self, canvas, axes, useblit=True, horizOn=False, vertOn=True,
-                 **lineprops):
+                 **props):
```
As above: Not a selector, other scenario.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -1970,12 +1976,13 @@ def __init__(self, ax, onselect, direction, minspan=0, useblit=False,
         # Setup handles
-        props = dict(color=rectprops.get('facecolor', 'r'))
-        props.update(cbook.normalize_kwargs(handle_props, Line2D._alias_map))
+        _handle_props = dict(color=props.get('facecolor', 'r'))
```
```suggestion
        handle_props_ = dict(color=props.get('facecolor', 'r'))
```
The standard notation for circumventing name clashes is a trailing underscore. Leading underscores are for "private" variables, which does not make sense for a local variable.

timhoffm replied:
Anyway, it's probably better to write this as 
```
        handle_props = {
            'color': props.get('facecolor', 'r'),
            **cbook.normalize_kwargs(handle_props, Line2D._alias_map)}
```

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -2511,27 +2542,31 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
-            props = dict(markeredgecolor='r')
+        if props is None:
+            _handle_props = dict(markeredgecolor='black')
         else:
-            props = dict(markeredgecolor=rectprops.get('edgecolor', 'r'))
-        props.update(cbook.normalize_kwargs(marker_props, Line2D._alias_map))
+            _handle_props = dict(
+                markeredgecolor=props.get('edgecolor', 'black')
+                )
+        _handle_props.update(cbook.normalize_kwargs(handle_props,
+                                                    Line2D._alias_map))
```
Whole block is 
```
        props = {
            'markeredgecolor': (props or {}).get('edgecolor', 'black'),
            **cbook.normalize_kwargs(handle_props, Line2D._alias_map)}
```

Um, and did you intentionally change the default from 'r'  to 'black'?

ericpre replied:
Yes, this is to match the default `edgecolor`.

QuLogic commented:
> An additional complication is that `RectangleSelector` / `EllipseSelector` takes `lineprops` or `rectprops` depending on `drawtype`. Either solution for that is bad (`drawtype` deciding which parameters are accepted and `drawtype` deciding which values of a `prop` parameter are supported).

We should not worry about `drawtype` in designing the desired configuration API, because it's deprecated already.

timhoffm commented:
> We should not worry about `drawtype` in designing the desired configuration API, because it's deprecated already.

Thanks for pointing this out, I have overlooked this. However, I don't think it changes the above assessment. We're still at `rect_props` for `RectangleSelector` and `EllipseSelector`. The name is awkward for the latter. Though, I would also be ok to leave one `lineprops` / `rectprops` parameter per Selector.

ericpre commented:
I think that I have addressed all comments. I have no idea why the doc build is failing and I can't see how it is related to this PR; the warning is:
```
/home/circleci/project/lib/matplotlib/projections/geo.py:docstring of matplotlib.projections.geo.GeoAxes.set_xlim:47: WARNING: py:obj reference target not found: get_ylim

```

QuLogic commented:
A rebase should fix that.

ericpre commented:
> A rebase should fix that.

Ok, thanks!

QuLogic commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,10 @@
+RectangleSelector and EllipseSelector
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``maxdist`` is deprecated, use ``grab_range`` argument instead
+``marker_props`` is deprecated, use ``handle_props`` argument instead
```
```suggestion
The *maxdist* argument is deprecated, use *grab_range* instead.
The *marker_props* argument is deprecated, use *handle_props* instead.
```

QuLogic commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,10 @@
+
+PolygonSelector
+~~~~~~~~~~~~~~~
+``vertex_select_radius`` is deprecated, use ``grab_range`` argument instead
+``markerprops`` is deprecated, use ``handle_props`` argument instead
```
```suggestion
The *vertex_select_radius* argument is deprecated, use *grab_range* instead.
The *markerprops* argument is deprecated, use *handle_props* instead.
```

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -2102,12 +2106,13 @@ def __init__(self, ax, onselect, direction, minspan=0, useblit=False,
-        props = dict(color=rectprops.get('facecolor', 'r'))
-        props.update(cbook.normalize_kwargs(handle_props, Line2D._alias_map))
+        handle_props = {
+            'color': (props or {}).get('facecolor', 'r'),
```
`props` can't be `None` due to line 2080?

ericpre replied:
Yes, indeed!

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -3142,8 +3199,11 @@ class PolygonSelector(_SelectorWidget):
     point.
     """
 
+    @_api.rename_parameter("3.5", "lineprops", "props")
+    @_api.rename_parameter("3.5", "markerprops", "handle_props")
+    @_api.rename_parameter("3.5", "vertex_select_radius", "grab_range")
     def __init__(self, ax, onselect, useblit=False,
-                 lineprops=None, markerprops=None, vertex_select_radius=15):
+                 props=None, handle_props=None, grab_range=10):
```
Why did the default change from 15?

ericpre replied:
This is to be consistent with the default value of the other selectors.

timhoffm commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,9 @@
+RectangleSelector and EllipseSelector
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The *maxdist* argument is deprecated, use *grab_range* instead.
```
`rectprops` -> `props` is missing. Same for all ->`props` renamings.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -2498,103 +2511,119 @@ def closest(self, x, y):
-        drawtype : {"box", "line", "none"}, default: "box"
-            Whether to draw the full rectangle box, the diagonal line of the
-            rectangle, or nothing at all.
+            ``dict(facecolor='red', edgecolor='black', alpha=0.2, fill=True)``
```
```suggestion
        ``dict(facecolor='red', edgecolor='black', alpha=0.2, fill=True)``
```

ericpre commented:
Thanks @timhoffm, I have updated the span selector example too.

timhoffm commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,19 @@
+
+SpanSelector
+~~~~~~~~~~~~
+The *rectprops* argument is deprecated, use *props* instead.
```
Missing `handle_grab_distance` -> `grab_range`

timhoffm commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,19 @@
+RectangleSelector and EllipseSelector
```
I would give this some context so that it does not appear like a random change of many parameters, e.g.

```
Unification of Selector API
~~~~~~~~~~~~~~~~~~~~~~~~~~~
The API for Selector widgets has been unified to use

- `props` for the properties of the Artist representing the selection.
- `handle_props` for the Artists representing handles for modifying the selection.
- `grab_range` for the maximal tolerance to grab a handle with the mouse.

This affects the following parameters and attributes:
```

+ Following section headers with 1 level reduced.

ericpre replied:
Done in 00444e4.

timhoffm commented on doc/api/next_api_changes/deprecations/20585-EP.rst:
```diff
@@ -0,0 +1,19 @@
+
+PolygonSelector
+~~~~~~~~~~~~~~~
+The *vertex_select_radius* argument is deprecated, use *grab_range* instead.
```
```suggestion
The *vertex_select_radius* argument and attribute is deprecated, use *grab_range* instead.
```
Same for `maxdist` and `handle_grab_distance`.

timhoffm commented on lib/matplotlib/widgets.py:
```diff
@@ -2685,6 +2714,10 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
 
     interactive = _api.deprecate_privatize_attribute("3.5")
 
+    maxdist = _api.deprecated("3.5")(
+        property(lambda self: self.grab_range)
+        )
```
I don't think this will work as intended. `deprecated()` is a decorator that must be applied to a function. However a property is not callable, so I think the warning will not be emitted. Did you check this?

Also you need a replacement for the setter.

Something like this *may* work, please verify. I did not have time to check myself.

```suggestion
    maxdist = property(
        _api.deprecated("3.5", alternative='grab_range')(lambda self: self.grab_range)
        _api.deprecated("3.5", alternative='grab_range')(lambda: self, val: self.grab_range = val)
        )
```

Ping @anntzer is there a better way to rename an attribute?

anntzer replied:
Actually `deprecated` has special handling for properties, and does work as is -- but you still need to provide a setter, so something like
```python
maxdist = _api.deprecated("3.5", name="maxdist", alternative="grab_range")(
    property(lambda self: self.grab_range,
             lambda self, value: setattr(self, "grab_range", value)))
```

ericpre replied:
I did check that the property was working but forgot about the setter! This should be fixed in 8682b5f now.

anntzer replied:
(side point: support for properties is explicitly listed at https://matplotlib.org/devdocs/api/_api_api.html#matplotlib._api.deprecation.deprecated)

timhoffm reviewed (commented):
Thanks for the quick response. I did a careful full review now, which revealed still some open issues.
